k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,382 @@
+ *
+ * @todo
+ *  Find a way to avoid this and make .h/.cc with cuda code
+ *  could be parsed by host compiler (specially, GNU-gcc).
+ *  (May assgin to nvcc to take control through change cmake
+ *  compiler and options. Then, it need cmake-3.18 `FindCUDAToolkit`
+ *  or other config helpers to make things easy.)
+ */
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+#include <device_launch_parameters.h>
+#include <driver_types.h>
+
+#include <stdio.h>
+#include <assert.h>
+
+#include "k2/csrc/cuda/arch.h"
+
+namespace k2 {
+
+/**
+ * @brief Define a macro `K2_MAKE_ERROR_CHECK` that turn on the macros bellow
+ *        to make errors get checked, if NDEBUG is undefined.
+ *
+ * @details If in NDEBUG mode, all macros here would do nothing.
+ */
+#if (!defined(NDEBUG) && !defined(K2_MAKE_ERROR_CHECK))
+#define K2_MAKE_ERROR_CHECK
+#endif
+
+/**
+ * @brief A static assertion
+ *
+ * @param[in] exp the compile-time boolean expression that must be true
+ * @param[in] msg an error message if exp is false
+ *
+ * @note `static_assert` is supported by both of host and device.
+ *
+ * @code{.cpp}
+ * K2_STATIC_ASSERT(DEFINED_SHAPE % DEFINED_X == 0);
+ * @endcode
+ */
+#ifndef K2_STATIC_ASSERT
+#define K2_STATIC_ASSERT(exp, msg) static_assert(exp, msg)
+#endif
+
+/**
+ * @brief Check if the expression is true.
+ *
+ * @details Implemented by `assert`, which is supported by host and device.
+ *
+ * @param[in] exp the boolean expression that should be true
+ *
+ * @note
+ *  `assert(exp)` in device, if exp == 0, the kernel excution is halted.
+ *  If the program is run within a debugger, this triggers a breakpoint and the
+ *  debugger can be used to inspect the current state of the device.
+ *  Otherwise, each thread for which expression is equal to zero prints a
+ *  message to stderr after synchronization with the host via
+ *  `cudaDeviceSynchronize(),cudaStreamSynchronize(),cudaEventSynchronize()`.
+ *  The format of this message is as follows:
+ *
+ *  @code
+ *  <filename>:<line number>:<function>:
+ *  block: [blockId.x,blockId.x,blockIdx.z],
+ *  thread: [threadIdx.x,threadIdx.y,threadIdx.z]
+ *  Assertion `<expression>` failed.
+ *  @endcode
+ *
+ *  @code{.cpp}
+ *  K2_ASSERT(1 == 1);
+ *  @endcode
+ */
+#ifndef K2_ASSERT
+#define K2_ASSERT(exp) assert(exp)
+#endif
+
+/**
+ * @brief Check if two arguments are equal.
+ *
+ * @details Implemented by `assert`, which is supported by host and device.
+ *
+ * @param[in] a left argument to compare
+ * @param[in] b right argument to compare
+ *
+ * @note
+ *  Assertions are for debugging purposes. They can affect performance
+ *  and it is therefore recommended to disable them in production code.
+ *  `assert` can be disabled at compile time by defining the `NDEBUG`
+ *  preprocessor macro before including assert.h.
+ *  Thus, `K2_MAKE_ERROR_CHECK` is ignored here.
+ *
+ *  @code{.cpp}
+ *  K2_CHECK_EQ(1, 1);
+ *  @endcode
+ */
+#ifndef K2_CHECK_EQ
+#define K2_CHECK_EQ(a, b) assert( a == b )
```

Comment:
Please **always** add parenthesis to your macro arguments.

```cpp
assert((a) == (b))
```

Reply:
So we do not need something like `K2_CHECK_EQ(a, b) << "some error message"` ?

Reply:
@songmeixu I'd rather have it print out the error message, including the values of a and b,  and line numbers, if it fails.  This will make debugging much easier.

Reply:
.. also of course there will be other similar macros like K2_CHECK, K2_CHECK_GE, K2_CHECK_LT and so on.

Reply:
@songmeixu there are some comments here you need to address: (1) using parentheses, (2) printing out the values of a and b and the line numbers if it fails.
I'm thinking that if we want a regular assert that just dies, we can use assert().

Reply:
Noticed that `assert` in cuda has printed some error messages, e.g. 
```
test.cu:19: void testAssert(): block: [0,0,0], thread: [0,0,0] Assertion `should_be_one` failed.
```
so maybe we need to avoid the duplicate of line numbers when printing error messages.